Adding NEURON simulator package

pramodskumbhar opened the pull request:
This is still based on `Package` and not on `AutotoolsPackage `. For cross-compiling environment we need to follow two step compilation process and hence it's bit complicated (but still doable). In it's current format it has been tested on all platforms (OS X, Linux, BG-Q, Cray, KNL) and hence submitting this in the current form.

adamjstewart commented on var/spack/repos/builtin/packages/neuron/package.py:
```diff
@@ -0,0 +1,240 @@
+
+    def url_for_version(self, version):
+        url = "http://www.neuron.yale.edu/ftp/neuron/versions/v{0}/nrn-{0}.tar.gz"
+        return url.format(version, version)
```
You shouldn't need this, it already works:
```
$ spack url parse http://www.neuron.yale.edu/ftp/neuron/versions/v7.4/nrn-7.4.tar.gz
==> Parsing URL: http://www.neuron.yale.edu/ftp/neuron/versions/v7.4/nrn-7.4.tar.gz

==> Matched version regex  0: r'^[a-zA-Z+._-]+[._-]v?(\\d[\\d._-]*)$'
==> Matched  name   regex  7: r'^([A-Za-z\\d+\\._-]+)$'

==> Detected:
    http://www.neuron.yale.edu/ftp/neuron/versions/v7.4/nrn-7.4.tar.gz
                                                        --- ~~~
    name:    nrn
    version: 7.4

==> Substituting version 9.9.9b:
    http://www.neuron.yale.edu/ftp/neuron/versions/v9.9.9b/nrn-9.9.9b.tar.gz
                                                           --- ~~~~~~
```

pramodskumbhar replied:
done!

adamjstewart commented on var/spack/repos/builtin/packages/neuron/package.py:
```diff
@@ -0,0 +1,240 @@
+            # TODO: on cray systems we get an error while linking even if
+            # we use cc wrapper. Hence add explicitly add mpi library
+            if spec.satisfies('+mpi'):
+                options.append('LIBS=-lmpich')
```
Does this work for all MPI libraries? Try `spec['mpi'].libs.link_flags`. See http://spack.readthedocs.io/en/latest/llnl.util.html#llnl.util.filesystem.LibraryList.link_flags for details.

pramodskumbhar replied:
done

adamjstewart commented on var/spack/repos/builtin/packages/neuron/package.py:
```diff
@@ -0,0 +1,240 @@
+        if spec.satisfies('+python'):
+            py_prefix = spec['python'].prefix
+            py_version = 'python{0}'.format(spec['python'].version.up_to(2))
+            python_exec = '%s/bin/%s' % (py_prefix, py_version)
```
Try:
```python
python_exec = spec['python'].command.path
```

pramodskumbhar replied:
done

adamjstewart commented on var/spack/repos/builtin/packages/neuron/package.py:
```diff
@@ -0,0 +1,240 @@
+                # PYINCDIR, PYLIB, PYLIBDIR etc
+
+                py_lib = spec['python'].prefix.lib
+                py_inc = '%s/include/%s' % (py_prefix, py_version)
```
Many of these things can be done more easily and reliably with Spack's `libs` and `headers` support. See http://spack.readthedocs.io/en/latest/llnl.util.html#module-llnl.util.filesystem for details. See the `boost` package for examples.

pramodskumbhar replied:
@adamjstewart : thanks for pointer!

In the boost it is used as `spec['python'].libs[0]`. But I am seeing an error:

```
==> Executing phase : 'install'
==> '/usr/local/Cellar/python3/3.6.2/Frameworks/Python.framework/Versions/3.6/bin/python3.6' '-c' 'from distutils.sysconfig import get_config_h_filename; print(get_config_h_filename())'
---> /usr/local/Cellar/python3/3.6.2/Frameworks/Python.framework/Versions/3.6/include/python3.6m
==> '/usr/local/Cellar/python3/3.6.2/Frameworks/Python.framework/Versions/3.6/bin/python3.6' '-c' 'from distutils.sysconfig import get_config_var; print(get_config_var('"'"'LIBDIR'"'"'))'
==> '/usr/local/Cellar/python3/3.6.2/Frameworks/Python.framework/Versions/3.6/bin/python3.6' '-c' 'from distutils.sysconfig import get_config_var; print(get_config_var('"'"'PYTHONFRAMEWORKPREFIX'"'"'))'
==> '/usr/local/Cellar/python3/3.6.2/Frameworks/Python.framework/Versions/3.6/bin/python3.6' '-c' 'from distutils.sysconfig import get_config_var; print(get_config_var('"'"'LIBRARY'"'"'))'
==> Error: RuntimeError: Unable to locate python libraries in /usr/local/opt/python3/Frameworks/Python.framework/Versions/3.6/lib
```

The `lib` folder has:

```
$ ll /usr/local/opt/python3/Frameworks/Python.framework/Versions/3.6/lib/
total 16
drwxr-xr-x    6 kumbhar  admin   204 Jul 17 05:45 .
drwxr-xr-x    9 kumbhar  admin   306 Jul 17 05:45 ..
lrwxr-xr-x    1 kumbhar  admin     9 Jul 17 05:45 libpython3.6.dylib -> ../Python
lrwxr-xr-x    1 kumbhar  admin     9 Jul 17 05:45 libpython3.6m.dylib -> ../Python
drwxr-xr-x    5 kumbhar  admin   170 Aug 14 20:41 pkgconfig
drwxr-xr-x  205 kumbhar  admin  6970 Aug 14 20:41 python3.6
```

Also if I have:
```
l = LibraryList(['/usr/local/Cellar/python3/3.6.2/Frameworks/Python.framework/Versions/3.6/lib/libpython3.6.dylib'])
```

I expected `l.ld_flags` to be `-lpython3.6` but get `-lpython3`:

```
-L/usr/local/Cellar/python3/3.6.2/Frameworks/Python.framework/Versions/3.6/lib -lpython3
```

Anything obvious thing I am missing?

adamjstewart replied:
This looks like a bug in our definition of `LibraryList.names`. Since libraries can contain suffixes like `libfoo.dylib.1.0.6`, we search for the first period and strip everything after. This doesn't work when the library name contains a period. Let me see if I can fix it.

adamjstewart replied:
Can you test out #5118?

adamjstewart commented on var/spack/repos/builtin/packages/neuron/package.py:
```diff
@@ -0,0 +1,240 @@
+        # need to add fpic and enabled-shared
+        if spec.satisfies('%pgi'):
+            options.extend(['CFLAGS=-fPIC %s' % flags,
+                            'CXXFLAGS=-fPIC %s' % flags,
```
I would prefer you use `self.compiler.pic_flag`. Although this section of code is only run for one compiler, I wouldn't be shocked if this flag changes in future releases (Intel's OpenMP flag changed).

pramodskumbhar replied:
done

adamjstewart commented on var/spack/repos/builtin/packages/neuron/package.py:
```diff
@@ -0,0 +1,240 @@
+                            'linux_nrnmech=no'])
+
+        # on os-x disable building carbon 'click' utility (deprecated)
+        if(sys.platform == 'darwin'):
```
Need space after `if`

pramodskumbhar replied:
done

adamjstewart commented on var/spack/repos/builtin/packages/neuron/package.py:
```diff
@@ -0,0 +1,240 @@
+
+        # on os-x disable building carbon 'click' utility (deprecated)
+        if(sys.platform == 'darwin'):
+            options.extend(['macdarwin=no'])
```
`append('')` instead of `extend([''])`

adamjstewart commented on var/spack/repos/builtin/packages/neuron/package.py:
```diff
@@ -0,0 +1,240 @@
+                   'CC=%s' % which("gcc"),
+                   'CXX=%s' % which("g++")]
+
+        configure = Executable(join_path(self.stage.source_path, 'configure'))
```
Shouldn't need this line.

pramodskumbhar replied:
we are in `build` directory and hence need path of configure

pramodskumbhar commented:
@adamjstewart : only remaining change your requested is use of `spec[python].libs` instead of manually way that I implemented. Would it be ok if I do it after we get some concrete idea about #5112 ?

adamjstewart commented on var/spack/repos/builtin/packages/neuron/package.py:
```diff
@@ -0,0 +1,236 @@
+    version('7.4', '2c0bbee8a9e55d60fa26336f4ab7acbf')
+    version('7.3', '993e539cb8bf102ca52e9fefd644ab61')
+    version('7.2', '5486709b6366add932e3a6d141c4f7ad')
+    version('develop', git=github, preferred=True)
```
The develop version should never be preferred, as it can't be checksummed.

pramodskumbhar replied:
ok, removed `preferred`

adamjstewart commented on var/spack/repos/builtin/packages/neuron/package.py:
```diff
@@ -0,0 +1,236 @@
+                            '--enable-shared'])
+        else:
+            options.extend(['CFLAGS=%s' % flags,
+                            'CXXFLAGS=%s' % flags])
```
The CFLAGS/CXXFLAGS logic is repeated. If you bring it out of the if-statement, you don't need to repeat it.

pramodskumbhar replied:
done

adamjstewart commented on var/spack/repos/builtin/packages/neuron/package.py:
```diff
@@ -0,0 +1,236 @@
+
+    variant('mpi',           default=True,  description='Enable MPI parallelism')
+    variant('python',        default=True,  description='Enable python')
+    variant('static',        default=True,  description='Build static libraries')
```
Most packages have a `shared` variant. `+shared` implies shared libraries, `~shared` implies static libraries. For comparison:
```
$ grep variant.*shared */package.py | wc -l
      53
$ grep variant.*static */package.py | wc -l
       2
```

pramodskumbhar replied:
done

adamjstewart commented on var/spack/repos/builtin/packages/neuron/package.py:
```diff
@@ -0,0 +1,236 @@
+                   '--with-nmodl-only',
+                   '--without-x',
+                   'CC=%s' % which("gcc"),
+                   'CXX=%s' % which("g++")]
```
Use `spack_cc` and `spack_cxx` (Python variables, don't need which). The logic you use here doesn't work if someone chooses any other compiler.

pramodskumbhar replied:
I have added Note/Todo to explain the situation : there is two-step build on cross-compiling architectures where we have to use front-end compiler (like gcc). Currently this is the only way I can think of getting this done (`spack_cc` will be pointing to back-end)

adamjstewart commented on var/spack/repos/builtin/packages/neuron/package.py:
```diff
@@ -0,0 +1,236 @@
+        # for bg-q we can't set XL as CC and CXX compiler
+        if 'bgq' in self.spec.architecture:
+            c_compiler = spec['mpi'].mpicc
+            cxx_compiler = spec['mpi'].mpicxx
```
This isn't going to work if someone tries to build `neuron~mpi`. You need to check whether or not `'+mpi' in spec`.

pramodskumbhar replied:
I have added `conflict` for `bg-q` and `~mpi` at the beginning. If you think it's good practice add check again, I can do.

adamjstewart replied:
It's always a good practice to double check. Someone might remove the conflict someday if they really want to build `~mpi` on `bg-q`. You should also be setting those the same way for `+mpi` on non-BGQ platforms.

pramodskumbhar replied:
added

adamjstewart commented:
I've never cross-compiled something before, so I would like to see what more experienced Spack developers have to say about this PR.

pramodskumbhar commented:
ok understood.

Just as an additional info: (as a developer of this neuron simulator), I have built this package on cross compiling architectures like bg-q,cray, knl architectures using Spack. On bg-q we build this with xlc and cray/knl with Intel/Cray/PGI. But even if we use gcc toolchain and `which gcc` gives back-end compiler then the components being built are minimal and won't have any backend specific instructions.

pramodskumbhar commented:
@adamjstewart : 

* I rebuild this on cross-compiling platforms using possible toolchains : cray, intel, pgi, xl compiler.  Use of `-march=native` for gcc makes front-end build robust even if selected GNU toolchain is for backend architecture.
* I am making `develop` as a preferred version because of `bug` in handling long path in NEURON (with spack, on typical hpc system the `stage` path is long e.g. ` /users/kumbhar/workarena/systems/daint/softwares/sources/spack/opt/spack/cray-CNL-haswell/intel-17.0.0.098/neuron-develop-otqxofckueva7zmjuionqpxvcoiq65ki/x86_64/` and I don't want user to blame `spack` for the build failure.

If there are no other objections, I am really interested getting this merged.

pramodskumbhar commented:
Removed previous hack of using `which("gcc")`. Now use architecture specific way to build components for front-end.

pramodskumbhar commented:
Suggestion accommodated from previous reviews:
 * removed stuff like `which(gcc)`
 * python configure args simplified
 * removed `preferred develop` with newly released tarball
 * Improved python 2 & 3 support with cross compilation

@adamjstewart : if you would like to improve anything, please let me know.

adamjstewart commented:
@pramodskumbhar I just started grad school and am too swamped to review and merge PRs for the foreseeable future. I'll let someone else handle this PR.

pramodskumbhar commented:
Oh ok, no problem @adamjstewart!

adamjstewart commented:
This PR is old and no one has complained about it in a while so I'm just going to merge it. If anyone encounters any problems, they can submit another PR to fix it.
